### ISSUE TITLE
Persistent connections don't persist

### ISSUE THREAD

**Issue by minenok-tutu** [mongodb/mongo-php-driver#274]
I turned on mongodb.debug and I see such output from an apache request:

```
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Connection string: 'mongodb://my.database.host:27017/database'
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Creating Manager, phongo-1.1.2[stable] - mongoc-1.3.2(1.3.2), libbson-1.3.2(1.3.2), php-5.4.15
[2016-03-31T09:42:49+00:00]     client: DEBUG   > Using custom stream initiator.
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Connecting to 'my.database.host:27017[mongodb://my.database.host:27017/database]'
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Created: RSRC#739 as 'my.database.host:27017[mongodb://my.database.host:27017/database]'
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Setting timeout to: 0
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Setting timeout to: 0
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Reading got: 4 wanted: 0
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Setting timeout to: 0
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Reading got: 376 wanted: 0
[2016-03-31T09:42:49+00:00] PHONGO-BSON: TRACE   > encoding document
[2016-03-31T09:42:49+00:00]    cluster: INFO    > SCRAM: authenticating "user" (step 1)
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Setting timeout to: 300000
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Setting timeout to: 300000
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Reading got: 4 wanted: 4
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Setting timeout to: 300000
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Reading got: 191 wanted: 191
[2016-03-31T09:42:49+00:00]    cluster: INFO    > SCRAM: authenticating "user" (step 2)
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Setting timeout to: 300000
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Setting timeout to: 300000
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Reading got: 4 wanted: 4
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Setting timeout to: 300000
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Reading got: 120 wanted: 120
[2016-03-31T09:42:49+00:00]    cluster: INFO    > SCRAM: authenticating "user" (step 3)
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Setting timeout to: 300000
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Setting timeout to: 300000
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Reading got: 4 wanted: 4
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Setting timeout to: 300000
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Reading got: 90 wanted: 90
[2016-03-31T09:42:49+00:00]    cluster: INFO    > SCRAM: "user" authenticated
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Setting timeout to: 300000
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Setting timeout to: 300000
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Reading got: 4 wanted: 4
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Setting timeout to: 300000
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Reading got: 32 wanted: 32
[2016-03-31T09:42:49+00:00] PHONGO-BSON: TRACE   > encoding document
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Setting timeout to: 300000
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Setting timeout to: 300000
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Reading got: 4 wanted: 4
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Setting timeout to: 300000
[2016-03-31T09:42:49+00:00]     PHONGO: DEBUG   > Reading got: 882 wanted: 882

// some other operations

[2016-03-31T09:42:51+00:00]     PHONGO: DEBUG   > Not destroying RSRC#-1
```

As I see in php_phongo.c:717 library really does not want to destroy an existing resource, because it wants it to persist. But at this method the process has the resource id -1, which says "oops, the stream has already been destroyed".

I could not figure out myself, how this can happen, hope you guys can.

**Comment by minenok-tutu** [comment]
UPD: I've updated to 1.1.5 and still facing this issue.

Hope this helps:
[2016-04-04T20:40:28+00:00]     cursor: TRACE   >  EXIT: mongoc_cursor_next():1000
[2016-04-04T20:40:28+00:00]     cursor: TRACE   > ENTRY: mongoc_cursor_error():923
[2016-04-04T20:40:28+00:00]     cursor: TRACE   > ENTRY: _mongoc_cursor_error():941
[2016-04-04T20:40:28+00:00]     cursor: TRACE   >  EXIT: _mongoc_cursor_error():954
[2016-04-04T20:40:28+00:00]     cursor: TRACE   >  EXIT: mongoc_cursor_error():933
[2016-04-04T20:40:28+00:00]     cursor: TRACE   > ENTRY: mongoc_cursor_destroy():242
[2016-04-04T20:40:28+00:00]     cursor: TRACE   > ENTRY: _mongoc_cursor_destroy():259
[2016-04-04T20:40:28+00:00]     cursor: TRACE   >  EXIT: _mongoc_cursor_destroy():296
[2016-04-04T20:40:28+00:00]     cursor: TRACE   >  EXIT: mongoc_cursor_destroy():252
[2016-04-04T20:40:28+00:00]     mongoc: TRACE   > ENTRY: mongoc_topology_description_destroy():89
[2016-04-04T20:40:28+00:00]     mongoc: TRACE   > ENTRY: mongoc_server_description_destroy():145
[2016-04-04T20:40:28+00:00]     mongoc: TRACE   >  EXIT: mongoc_server_description_destroy():151
[2016-04-04T20:40:28+00:00]     mongoc: TRACE   >  EXIT: mongoc_topology_description_destroy():103
[2016-04-04T20:40:28+00:00]     stream: TRACE   > ENTRY: mongoc_stream_destroy():103
[2016-04-04T20:40:28+00:00]     PHONGO: DEBUG   > Not destroying RSRC#-1

**Comment by jmikola** [comment]
> As I see in php_phongo.c:717 library really does not want to destroy an existing resource, because it wants it to persist. But at this method the process has the resource id -1, which says "oops, the stream has already been destroyed".

The current "Not destroying" log message is only printed when the PHP stream *does* exist and we want to leave it for the next request. I'm not sure why it has a resource ID of -1 (and I agree it's odd), but that is coming directly the the PHP stream struct. Do you have a reference for -1 signifying that the stream is destroyed? If possible, it's be worth [printing the state of that struct in GDB](http://stackoverflow.com/a/12618396/162228).

More importantly, is there any evidence of the socket actually being closed after each PHP request and not persisting? I expect you could cross-reference the script execution with MongoDB server logs to verify that.

**Comment by minenok-tutu** [comment]
> The current "Not destroying" log message is only printed when the PHP stream does exist and we want to leave it for the next request. I'm not sure why it has a resource ID of -1 (and I agree it's odd), but that is coming directly the the PHP stream struct. Do you have a reference for -1 signifying that the stream is destroyed?

The only related post I've found is this: http://stackoverflow.com/questions/6944300/can-php-return-a-resource-with-id-0
I think that "default value" is something could represent a destroyed stream.
Plus, as you can see at the log above, process gets a positive resource ID when connecting to MongoDB.

> If possible, it's be worth printing the state of that struct in GDB.

I don't know how to debug PHP process, unless it is failed with segfault. Can you show me how? :)

> More importantly, is there any evidence of the socket actually being closed after each PHP request and not persisting? I expect you could cross-reference the script execution with MongoDB server logs to verify that.

We have some utility, which became a daemon one day. LA on the MongoDB server dropped 10 times, and the stats metric called "command". The most probable cause: we had many processes authenticating all the time, and now we don't.


And I see some weird thing here: https://github.com/mongodb/mongo-c-driver/blob/master/src/mongoc/mongoc-cursor.c#L293
Can you tell me, how to avoid it?


Did you tested persistent connections? Your log is not the same?

**Comment by jmikola** [comment]
> And I see some weird thing here: https://github.com/mongodb/mongo-c-driver/blob/master/src/mongoc/mongoc-cursor.c#L293
Can you tell me, how to avoid it?

That bit of code is only relevant if the cursor is in exhaust mode. Are you using the "exhaust" option in [`MongoDB\Driver\Query::__construct()`](http://php.net/manual/en/mongodb-driver-query.construct.php)?

> Did you tested persistent connections? Your log is not the same?

I've not reproduced anything here. If you can share a stand-alone script that reproduces the issue and log messages, I'd appreciate it. [PHP's built-in web server](http://php.net/manual/en/features.commandline.webserver.php) would be the preferred web SAPI to demonstrate this (unless the bug is due to a particular web SAPI). The built-in web server can persist resources between requests, so it should demonstrate the same behavior. It's also significantly easier to start up than Apache and its single-thread nature also ensures that multiple requests will all hit the same process (equivalent to running Apache or nginx with a worker pool of one).

**Comment by minenok-tutu** [comment]
Seems like GC collects the connection at some point.
Storing the connection at private property of wrapping class -> RSRC#-1
Storing at the $GLOBALS variable -> RSRC#<some positive int>

It's very weird. Anyway, seems like persistent connections aren't useful at all. I've tested the fix with $GLOBALS on production environment, and didn't get any significant changes on MongoDB instance. I think, persistent connections do not persist huge amount of time and/or Apache have very large connection pool.

It's sad, but thank for helping!

**Comment by jmikola** [comment]
> I've tested the fix with $GLOBALS on production environment, and didn't get any significant load changes on MongoDB instance. I think, persistent connections do not persist huge amount of time and/or Apache have very large connection pool.

The persistent connections should last as long as the PHP worker. Earlier, I asked if there was "any evidence of the socket actually being closed after each PHP request and not persisting" and suggested consulting the MongoDB server logs to verify. I didn't receive any response from you on that.

I also asked about whether you were using the "exhaust" option in [`MongoDB\Driver\Query::__construct()`](http://php.net/manual/en/mongodb-driver-query.construct.php) (due to the libmongoc code you referenced). Based on the trace you also shared, that is the only reason I can find for the connection being destroyed and not persisted (due to being considered a failure); however, in that case I would expect the driver to log "Destroying RSRC" from `phongo_stream_failed()` in its trace.

As previously asked, a script reproducing the exact issue you're seeing would be essential to investigating this. [PHP's built-in web server](http://php.net/manual/en/features.commandline.webserver.php) should be suitable, since its connection persistence behavior should be identical to what we'd have in Apache or nginx/FPM.

**Comment by minenok-tutu** [comment]
> The persistent connections should last as long as the PHP worker. Earlier, I asked if there was "any evidence of the socket actually being closed after each PHP request and not persisting" and suggested consulting the MongoDB server logs to verify. I didn't receive any response from you on that.

I'm sorry, right now I can't get that. I'll show them later.

> I also asked about whether you were using the "exhaust" option in MongoDB\Driver\Query::__construct() (due to the libmongoc code you referenced). Based on the trace you also shared, that is the only reason I can find for the connection being destroyed and not persisted (due to being considered a failure); however, in that case I would expect the driver to log "Destroying RSRC" from phongo_stream_failed() in its trace.
No, exhaust option is not used.

> As previously asked, a script reproducing the exact issue you're seeing would be essential to investigating this. PHP's built-in web server should be suitable, since its connection persistence behavior should be identical to what we'd have in Apache or nginx/FPM.
None of the simple scripts I've written are showing the issue. It only happens inside the big application AND if I use private arrays to store the connection inside this deep-down-connector-wrapper.

**Comment by franek** [comment]
I have tried to migrate a big application to PHP7 with pecl/mongodb and it seems that the new driver does not support persistence connexion.

```
<?php
// t.php
$filter = array();
$options = array('limit' => 5);

$query = new MongoDB\Driver\Query($filter, $options);

$manager = new MongoDB\Driver\Manager(
    "mongodb://localhost:27017",
    ['username' => 'user', 'password' => 'user', 'db' => 'database', 'authSource' => 'database']
);
$readPreference = new MongoDB\Driver\ReadPreference(MongoDB\Driver\ReadPreference::RP_PRIMARY);
$cursor = $manager->executeQuery("database.my-collection", $query, $readPreference);

foreach($cursor as $document) {
    var_dump($document->item);
}
```

I have started php with the built-in server : 
```
$ php -S localhost:8000 t.php
```
and run a batch of request : 
```
$ ab -c 10 -n 10 localhost:8000/t.php
```
Into mongod.log : 
```
2016-06-08T11:47:43.951+0200 I ACCESS   [conn34908] Successfully authenticated as principal user on database
2016-06-08T11:47:43.997+0200 I ACCESS   [conn34908] Successfully authenticated as principal user on database
2016-06-08T11:47:44.035+0200 I ACCESS   [conn34908] Successfully authenticated as principal user on database
2016-06-08T11:47:44.065+0200 I ACCESS   [conn34908] Successfully authenticated as principal user on database
2016-06-08T11:47:44.102+0200 I ACCESS   [conn34908] Successfully authenticated as principal user on database
2016-06-08T11:47:44.134+0200 I ACCESS   [conn34908] Successfully authenticated as principal user on database
2016-06-08T11:47:44.173+0200 I ACCESS   [conn34908] Successfully authenticated as principal user on database
2016-06-08T11:47:44.208+0200 I ACCESS   [conn34908] Successfully authenticated as principal user on database
2016-06-08T11:47:44.243+0200 I ACCESS   [conn34908] Successfully authenticated as principal user on database
2016-06-08T11:47:44.273+0200 I ACCESS   [conn34908] Successfully authenticated as principal opa on database
```

environment : 
```
$ php -v
PHP 7.0.7 (cli) (built: May 30 2016 09:43:45) ( NTS )
Copyright (c) 1997-2016 The PHP Group
Zend Engine v3.0.0, Copyright (c) 1998-2016 Zend Technologies
    with Zend OPcache v7.0.6-dev, Copyright (c) 1999-2016, by Zend Technologies
    with blackfire v1.10.6, https://blackfire.io, by Blackfireio Inc.
$ php --ri mongodb
mongodb support => enabled
mongodb version => 1.1.6
mongodb stability => stable
libmongoc version => 1.3.5
libbson version => 1.3.5

Directive => Local Value => Master Value
mongodb.debug => no value => no value
```
I will try with pecl/mongodb-1.1.7

Can you confirm us the problem ? 

Thanks,

**Comment by derickr** [comment]
@franek All the connections in the log show the same connection ID. It only tells you about authentication on that connection. That does not indicate that the connection is not persistent.

**Comment by franek** [comment]
@derickr ok. Thanks for your answer. How can we be sure that the connection is persisted ? Since we have migrated to PHP7-pecl/mongodb, we have many more entries in our logs like this one "Successfully authenticated as...". We did not have these entries with the legacy driver (PHP5). How do you explain them ?

**Comment by jmikola** [comment]
@franek: Connection persistent in 1.1.x is currently handled by [`phongo_stream_initiator()`](https://github.com/mongodb/mongo-php-driver/blob/1.1.7/php_phongo.c#L1126), which is simply responsible for opening the socket (or returning an existing, persisted socket). I imagine that libmongoc is still authenticating on its own when that socket is used in a new request, since libmongoc's client has no recollection of the socket's state. For the same reason, subsequent PHP requests are prone to doing topology discover (i.e. issuing `isMaster` commands).

This should be addressed by [PHPC-605](https://jira.mongodb.org/browse/PHPC-605) and [PHPC-433](https://jira.mongodb.org/browse/PHPC-433) in 1.2.0, where we plan to remove `phongo_stream_initiator()` in favor of simply keeping the libmongoc client alive between requests. @derickr has already implemented such logic in the HHVM driver with [HHVM-123](https://jira.mongodb.org/browse/HHVM-123).

**Comment by isegal** [comment]
Confirmed and reproducible with php7.
Here is a client-side packet capture of two requests:
https://gist.github.com/isegal/00423f43de267634348a431dadab8714

We have two mongo replicas and an arbiter. It appears to query each with isMaster().

To resolve the latter problem I've seen some clients such as an earlier version of mongofill rely on rs.status() instead in order to minimize the amount of polls.

A couple of advantages we see in production:
1. Less tcp traffic to servers.
2. If any of the servers (other than the one we initially connect to) are down, the client won't attempt to connect and/or wait to establish connection for the purpose of isMaster() call.


On a side note, the lack of connection persistence seems to be a current state of things in HHVM when using pfsockopen() so mongofill does not persist either.
We think it's related to garbage collection:
https://github.com/facebook/hhvm/issues/6631

**Comment by jmikola** [comment]
> We have two mongo replicas and an arbiter. It appears to query each with isMaster().

@isegal: If I understand correctly, you do see the connections persisted and the issue is just that each PHP request starts with a round of isMaster calls. Is that correct? If so, that's exactly what we're addressing in [PHPC-433](https://jira.mongodb.org/browse/PHPC-433), which has since been merged to master (https://github.com/mongodb/mongo-php-driver/pull/376) for inclusion with 1.2.0.

**Comment by isegal** [comment]
Checked again, actually yes we do see persisting connections. It looks like we are indeed seeing PHPC-433. Will wait for 1.2.0, thank you.

**Comment by jmikola** [comment]
[1.2.0alpha2](https://github.com/mongodb/mongo-php-driver/releases/tag/1.2.0alpha2) has been released, which includes significant changes to our connection handling and persistence. We've entirely dropped PHP's streams API and now rely on libmongoc for all socket communication, which I expect may resolve a number of outstanding connection handling and timeout issues.

@minenok-tutu, @franek, @isegal: Please attempt reproducing your original problems with 1.2.0alpha2 and let us know if the issue persists. Thanks!

**Comment by jmikola** [comment]
Note: we've just released [1.1.9](https://github.com/mongodb/mongo-php-driver/releases/tag/1.1.9), which includes a back-ported fix (in libmongoc) for spurious timeouts ""No suitable servers found" exceptions when one or more replica sets nodes were inaccessible. The bug was specific to the PHP stream API integration, so it does not apply to version [1.2.0alpha2](https://github.com/mongodb/mongo-php-driver/releases/tag/1.2.0alpha2) and later.

Since this issue has been sitting for some time without confirmation, I'm going to close it out. Feel free to follow-up here or open a new issue as needed.

**Comment by franek** [comment]
@jmikola sorry for the delay ;) Problem is solved on my side with 1.2.2 (I didn't try 1.2.0 or 1.2.1). :+1:
